Add assertions on ios_base::setf to prevent invalid flag combination.

plming opened the pull request:
## Explanation
Calling `ios_base::setf` method can lead to invalid combinations on `fmtflags` bitflag.

I know there's way to use `setf(flag, mask)` but below code doesn't omit any errors, so user can misuse it.

This PR adds some `STL_ASSERT` on `setf` to prevent invalid flag status.

## Before changes
```cpp
cout.setf(ios_base::dec);
cout.setf(ios_base::hex); // No compiler or runtime error

cout << 42 << endl; // Outputs 42, but intent is unclear
```

## After changes
```cpp
cout.setf(ios_base::dec);
cout.setf(ios_base::hex); // Debug assertion failed

cout << 42 << endl;
```

AlexGuteniev commented on stl/inc/xiosbase:
```diff
@@ -248,6 +248,11 @@ public:
         const ios_base::fmtflags _Oldfmtflags = _Fmtfl;
         _Fmtfl |= _Newfmtflags & _Fmtmask;
+
+        _STL_ASSERT(__popcnt(_Fmtfl & basefield) < 2, "Multiple basefield flags set. Try setf(flags, mask) instead.");
```
`__popcnt` intrinsic will emit `popcnt` instruction, without fallback path. We still support CPUs without this instruction by the STL, even though they may not be supported by the latest Windows.

To have `popcount` with fallback path and available in pre-C++20 mode, use `_Checked_popcount` from `<__msvc_bit_utils.hpp>`.

frederick-vs-ja commented:
I don't think we can make such assertions _just_ because the intent is unclear. It was clarified in LWG-287, which is NAD, that nonsense combinations still cause well-defined behavior.

StephanTLavavej commented:
I agree with @frederick-vs-ja's comment https://github.com/microsoft/STL/pull/5503#issuecomment-2878755458 - I don't see any preconditions in the Standard that forbid this.

I don't think we can merge this change. Am I missing some Standardese somewhere?

plming commented:
Thanks for feedbacks, I found `std::hexfloat` uses `fixed | scientific` flag, so I dropped assert about it.
And changed to `_Checked_popcount` for `popcnt` unsupported platforms.

But I still think that other fields(basefield, adjustfield)'s combinations are non-sense and not used from other codes I'm aware of.
Although there's no standards that forbid those combinations, I think we can defensively check on debug builds.
Still if such checks are inappropriate for STL, I'll accept and happy to withdraw the PR.

StephanTLavavej commented:
> Although there's no standards that forbid those combinations, I think we can defensively check on debug builds.

Library debug checks aren't like compiler warnings. The compiler can warn about anything it wants to, including valid but suspicious/likely-unintentional/risky code, because that happens at compile time and there's robust infrastructure for suppressing warnings. On the other hand, library debug checks happen at runtime, so we can only check for Standard-mandated precondition violations.

Thanks for looking into this. I'm going to close this PR as we can't merge it.
